File: Auto/car_app_library/navigation/src/main/java/com/google/android/libraries/car/app/samples/navigation/model/Instruction.java
/*
 * Copyright (C) 2020 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.android.libraries.car.app.samples.navigation.model;

import androidx.annotation.Nullable;
import com.google.android.libraries.car.app.model.Distance;
import com.google.android.libraries.car.app.navigation.model.Destination;
import com.google.android.libraries.car.app.navigation.model.Step;
import com.google.android.libraries.car.app.navigation.model.TravelEstimate;

/**
 * Container for scripting navigation instructions.
 *
 * <p>For example things which are triggered based on a users location changing can be put into the
 * script to simulate driving without having to provide a full location mockup.
 *
 * <p>In addition to a {@link Type}, each instruction has a length. All other parameters are
 * optional and related directly to which instruction is specified.
 */
public class Instruction {

  public enum Type {
    START_NAVIGATION,
    END_NAVIGATION,
    ADD_DESTINATION_NAVIGATION,
    POP_DESTINATION_NAVIGATION,
    ADD_STEP_NAVIGATION,
    POP_STEP_NAVIGATION,
    SET_TRIP_POSITION_NAVIGATION,
  };

  private final Type mType;
  private final long mLengthMillis;

  // Only support a single destination at the moment.
  @Nullable private final Destination mDestination;

  // Only support setting a single step.
  @Nullable private final Step mStep;
  @Nullable private final Distance mStepRemainingDistance;
  @Nullable private final TravelEstimate mStepTravelEstimate;
  @Nullable private final TravelEstimate mDestinationTravelEstimate;
  @Nullable private final String mRoad;

  /** Constructs a new builder of {@link Instruction}. */
  public static Builder builder(Type type, long lengthMs) {
    return new Builder(type, lengthMs);
  }

  public long getLengthMillis() {
    return mLengthMillis;
  }

  public Type getType() {
    return mType;
  }

  @Nullable
  public Destination getDestination() {
    return mDestination;
  }

  @Nullable
  public Step getStep() {
    return mStep;
  }

  @Nullable
  public Distance getStepRemainingDistance() {
    return mStepRemainingDistance;
  }

  @Nullable
  public TravelEstimate getStepTravelEstimate() {
    return mStepTravelEstimate;
  }

  @Nullable
  public TravelEstimate getDestinationTravelEstimate() {
    return mDestinationTravelEstimate;
  }

  @Nullable
  public String getRoad() {
    return mRoad;
  }

  private Instruction(Builder builder) {
    mType = builder.mType;
    mLengthMillis = builder.mLengthMs;
    mDestination = builder.mDestination;
    mStep = builder.mStep;
    mStepRemainingDistance = builder.mStepRemainingDistance;
    mStepTravelEstimate = builder.mStepTravelEstimate;
    mDestinationTravelEstimate = builder.mDestinationTravelEstimate;
    mRoad = builder.mRoad;
  }

  /** Builder for creating an {@link Instruction}. */
  public static final class Builder {
    private Type mType;
    private long mLengthMs;
    @Nullable private Destination mDestination;
    @Nullable private Step mStep;
    @Nullable private Distance mStepRemainingDistance;
    @Nullable private TravelEstimate mStepTravelEstimate;
    @Nullable private TravelEstimate mDestinationTravelEstimate;
    @Nullable private String mRoad;

    public Builder(Type type, long lengthMs) {
      mType = type;
      mLengthMs = lengthMs;
    }

    Builder setDestination(@Nullable Destination destination) {
      mDestination = destination;
      return this;
    }

    Builder setStep(@Nullable Step step) {
      mStep = step;
      return this;
    }

    Builder setStepRemainingDistance(@Nullable Distance stepRemainingDistance) {
      mStepRemainingDistance = stepRemainingDistance;
      return this;
    }

    Builder setStepTravelEstimate(@Nullable TravelEstimate stepTravelEstimate) {
      mStepTravelEstimate = stepTravelEstimate;
      return this;
    }

    Builder setDestinationTravelEstimate(@Nullable TravelEstimate destinationTravelEstimate) {
      mDestinationTravelEstimate = destinationTravelEstimate;
      return this;
    }

    Builder setRoad(@Nullable String road) {
      mRoad = road;
      return this;
    }

    /** Constructs the {@link Instruction} defined by this builder. */
    public Instruction build() {
      return new Instruction(this);
    }
  }
}

File: car_app_library/navigation/shared/src/main/java/androidx/car/app/sample/navigation/common/model/Instruction.kt
/*
 * Copyright (C) 2021 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package androidx.car.app.sample.navigation.common.model

import androidx.car.app.model.CarIcon
import androidx.car.app.model.Distance
import androidx.car.app.navigation.model.Destination
import androidx.car.app.navigation.model.Step
import androidx.car.app.navigation.model.TravelEstimate
import androidx.car.app.sample.navigation.common.model.Instruction.Type

/**
 * Container for scripting navigation instructions.
 *
 *
 * Each [Instruction] represents a change (delta) in the state of navigation corresponding
 * to a change in state that could occur during a real navigation session. In general a script
 * follows the following rough outline similar to a driving sessions:
 *
 *
 *  * Start Navigation
 *  * Add one or more destinations (as would be selected by a driver)
 *  * Add one or more steps (as would be provided by a routing algorithm)
 *  * Add updated positions until the next step is reached.
 *  * pop a step and send positions until the next step is reached.
 *  * Repeat until all steps are popped.
 *  * End Navigation.
 *
 *
 *
 * In addition to a [Type], each instruction specifies the duration until the next
 * instruction should be executed. A duration of zero allows multiple instructions to be stacked up
 * as if they were a single instruction. All other parameters are optional and related directly to
 * which instruction is specified.
 */
class Instruction internal constructor(builder: Builder) {
    enum class Type {
        START_NAVIGATION,
        END_NAVIGATION,
        ADD_DESTINATION_NAVIGATION,
        POP_DESTINATION_NAVIGATION,
        ADD_STEP_NAVIGATION,
        POP_STEP_NAVIGATION,
        SET_REROUTING,
        SET_ARRIVED,
        SET_TRIP_POSITION_NAVIGATION,
    }

    val type: Type = builder.mType
    val durationMillis: Long = builder.mDurationMillis

    // Only support a single destination at the moment.
    val destination: Destination? = builder.mDestination

    // Only support setting a single step.
    val step: Step? = builder.mStep
    val stepRemainingDistance: Distance? = builder.mStepRemainingDistance
    val stepTravelEstimate: TravelEstimate? = builder.mStepTravelEstimate
    val destinationTravelEstimate: TravelEstimate? = builder.mDestinationTravelEstimate
    val road: String? = builder.mRoad
    val shouldShowNextStep: Boolean = builder.mShouldShowNextStep
    val shouldShowLanes: Boolean = builder.mShouldShowLanes
    val junctionImage: CarIcon? = builder.mJunctionImage

    val notificationTitle: String? = builder.mNotificationTitle
    val notificationContent: String? = builder.mNotificationContent
    val notificationIcon: Int = builder.mNotificationIcon
    val shouldNotify: Boolean = builder.mShouldNotify

    /** Builder for creating an [Instruction].  */
    class Builder(var mType: Type, var mDurationMillis: Long) {
        var mDestination: Destination? = null
        var mStep: Step? = null
        var mStepRemainingDistance: Distance? = null
        var mStepTravelEstimate: TravelEstimate? = null
        var mDestinationTravelEstimate: TravelEstimate? = null
        var mRoad: String? = null
        var mShouldShowNextStep: Boolean = false
        var mShouldShowLanes: Boolean = false
        var mJunctionImage: CarIcon? = null

        var mNotificationTitle: String? = null
        var mNotificationContent: String? = null
        var mNotificationIcon: Int = 0
        var mShouldNotify: Boolean = false

        fun setDestination(destination: Destination?): Builder {
            mDestination = destination
            return this
        }

        fun setStep(step: Step?): Builder {
            mStep = step
            return this
        }

        fun setStepRemainingDistance(stepRemainingDistance: Distance?): Builder {
            mStepRemainingDistance = stepRemainingDistance
            return this
        }

        fun setStepTravelEstimate(stepTravelEstimate: TravelEstimate?): Builder {
            mStepTravelEstimate = stepTravelEstimate
            return this
        }

        fun setDestinationTravelEstimate(destinationTravelEstimate: TravelEstimate?): Builder {
            mDestinationTravelEstimate = destinationTravelEstimate
            return this
        }

        fun setRoad(road: String?): Builder {
            mRoad = road
            return this
        }

        fun setShouldShowNextStep(shouldShowNextStep: Boolean): Builder {
            mShouldShowNextStep = shouldShowNextStep
            return this
        }

        fun setShouldShowLanes(shouldShowLanes: Boolean): Builder {
            mShouldShowLanes = shouldShowLanes
            return this
        }

        fun setJunctionImage(junctionImage: CarIcon?): Builder {
            mJunctionImage = junctionImage
            return this
        }

        fun setNotification(
            shouldNotify: Boolean,
            notificationTitle: String?,
            notificationContent: String?,
            notificationIcon: Int
        ): Builder {
            mNotificationTitle = notificationTitle
            mNotificationContent = notificationContent
            mShouldNotify = shouldNotify
            mNotificationIcon = notificationIcon
            return this
        }

        /** Constructs the [Instruction] defined by this builder.  */
        fun build(): Instruction {
            return Instruction(this)
        }
    }

    companion object {
        /** Constructs a new builder of [Instruction].  */
        fun builder(type: Type, lengthMs: Long): Builder {
            return Builder(type, lengthMs)
        }
    }
}
